File: submissions/tessoku-book/tessoku_book_q.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;
int main() {
    // 入力
    int N;
    cin >> N;
    vector<int> A(N + 1), B(N + 1);
    for (int i = 2; i <= N; i++) {
        cin >> A[i];
    }
    for (int i = 3; i <= N; i++) {
        cin >> B[i];
    }
    // 動的計画法
    vector<int> dp(N + 1);
    dp[1] = 0;
    dp[2] = A[2];
    for (int i = 3; i <= N; i++) {
        dp[i] = min(dp[i - 1] + A[i], dp[i - 2] + B[i]);
    }
    // 最短経路を復元
    vector<int> ans;
    int place = N;
    while (true) {
        ans.push_back(place);
        if (place == 1) break;
        if (dp[place - 1] + A[place] == dp[place]) {
            place = place - 1;
        } else {
            place = place - 2;
        }
    }
    reverse(ans.begin(), ans.end());
    // 答えを出力
    cout << ans.size() << endl;
    for (int i = 0; i < ans.size(); i++) {
        if (i >= 1) {
            cout << ' ';
        }
        cout << ans[i];
    }
    cout << endl;
}

File: submissions/tessoku-book/tessoku_book_q.rs
use proconio::{fastout, input};
use std::cmp::min;
#[fastout]
fn main() {
    input! {
        n: usize,
        a: [usize; n - 1],
        b: [usize; n - 2],
    }
    let (k, p) = solve(n, &a, &b);
    println!("{}\n{}", k, p.iter().map(std::string::ToString::to_string).collect::<Vec<_>>().join(" "));
}
fn solve(n: usize, a: &Vec<usize>, b: &Vec<usize>) -> (usize, Vec<usize>) {
    let mut dp = vec![0; n + 1];
    dp[2] = a[0];
    for i in 3..dp.len() {
        dp[i] = min(dp[i - 1] + a[i - 2], dp[i - 2] + b[i - 3]);
    }
    let mut place = n;
    let mut p = Vec::new();
    loop {
        p.push(place);
        if place == 1 {
            break;
        }
        if dp[place - 1] + a[place - 2] == dp[place] {
            place -= 1;
        }
        else {
            place -= 2;
        }
    }
    p.reverse();
    (p.len(), p)
}